cpp/src/hdbscan/detail/soft_clustering.cuh

```diff
@@ -369,6 +496,143 @@ void all_points_membership_vectors(const raft::handle_t& handle,
+                       const value_t* X,
+                       const value_t* points_to_predict,
+                       size_t n_prediction_points,
+                       raft::distance::DistanceType metric,
+                       int min_samples,
+                       value_t* membership_vec)
+{
+  RAFT_EXPECTS(metric == raft::distance::DistanceType::L2SqrtExpanded,
+               "Currently only L2 expanded distance is supported");
+
+  auto stream      = handle.get_stream();
+  auto exec_policy = handle.get_thrust_policy();
+
+  size_t m                       = prediction_data.n_rows;
+  size_t n                       = prediction_data.n_cols;
+  value_idx n_selected_clusters  = prediction_data.get_n_selected_clusters();
+  value_t* deaths                = prediction_data.get_deaths();
+  value_idx* selected_clusters   = prediction_data.get_selected_clusters();
+  value_idx* index_into_children = prediction_data.get_index_into_children();
+  value_idx n_exemplars          = prediction_data.get_n_exemplars();
+  value_t* lambdas               = condensed_tree.get_lambdas();
+
+  auto counting = thrust::make_counting_iterator<value_idx>(0);
+  rmm::device_uvector<value_t> dist_membership_vec(n_prediction_points * n_selected_clusters,
+                                                   stream);
+
+  dist_membership_vector(handle,
+                         X,
+                         points_to_predict,
+                         n_prediction_points,
+                         n,
+                         n_exemplars,
+                         n_selected_clusters,
+                         prediction_data.get_exemplar_idx(),
+                         prediction_data.get_exemplar_label_offsets(),
+                         dist_membership_vec.data(),
+                         raft::distance::DistanceType::L2SqrtExpanded);
+
+  rmm::device_uvector<value_t> prediction_lambdas(n_prediction_points, stream);
+  rmm::device_uvector<value_idx> min_mr_inds(n_prediction_points, stream);
+
+  _compute_knn_and_nearest_neighbor(handle,
+                                    prediction_data,
+                                    X,
+                                    points_to_predict,
+                                    min_samples,
+                                    n_prediction_points,
+                                    min_mr_inds.data(),
+                                    prediction_lambdas.data(),
+                                    metric);
+
+  thrust::for_each(exec_policy,
+                   counting,
+                   counting + n_prediction_points,
+                   [lambdas,
+                    index_into_children,
+                    min_mr_inds        = min_mr_inds.data(),
+                    prediction_lambdas = prediction_lambdas.data()] __device__(auto idx) {
+                     value_t neighbor_lambda = lambdas[index_into_children[min_mr_inds[idx]]];
+                     if (neighbor_lambda < prediction_lambdas[idx])
```

Comment:
Another candidate for `raft::matrix::map_offset`. Please use raft functions wherever possible so we can centralize these optimizations and optimize them as needed.